### PR TITLE
chore: expose dropped metrics 

### LINKED ISSUE #1861: emit received and dropped count metrics
# Summary

Today, we do not have information on how many messages have been received and dropped.

# Use Cases

While debugging, why are messages not moving between the vertices?

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by samhith-kakarla**
Closes #1861

### CODE REVIEW

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -165,6 +165,12 @@ var (
 		Name:      "udf_write_total",
 		Help:      "Total number of Messages Written by UDF",
 	}, []string{LabelVertex, LabelPipeline, LabelVertexType, LabelVertexReplicaIndex, LabelPartitionName})
+
+	DroppedMessageCounter = promauto.NewCounterVec(prometheus.CounterOpts{
```
Comment (KeranYang) [inline pkg/metrics/metrics.go:169]: Do we already have `DropMessagesCount` in the same file?

Comment (whynowy) [reply pkg/metrics/metrics.go:169]: That one is mainly for dropping on buffer full...

use `ud_drop_total`?

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -165,6 +165,12 @@ var (
...
+		Subsystem: "forwarder",
+		Name:      "drop_counter",
+		Help:      "Total number of Messages Dropped",
```
Comment (whynowy) [inline pkg/metrics/metrics.go]: Total messages dropped by the user

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -165,6 +165,12 @@ var (
...
+	}, []string{LabelVertex})
```
Comment (whynowy) [inline pkg/metrics/metrics.go:173]: It should have `LabelVertex, LabelPipeline, LabelVertexType, LabelVertexReplicaIndex, LabelPartitionName`

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (whynowy) [reply pkg/metrics/metrics.go:169]: or `ud_dropped_messages_total`

File: pkg/sources/source.go [hunk 4]
```diff
@@ -400,6 +401,13 @@ func (sp *SourceProcessor) getTransformerGoWhereDecider(shuffleFuncMap map[strin
 
 		// Drop message if it contains the special tag
 		if sharedutil.StringSliceContains(tags, dfv1.MessageTagDrop) {
+			metrics.DroppedMessageCounter.With(map[string]string{
+				metrics.LabelVertex:             sp.VertexInstance.Vertex.Name,
+				metrics.LabelPipeline:           sp.VertexInstance.Vertex.Spec.PipelineName,
+				metrics.LabelVertexType:         string(dfv1.VertexTypeReduceUDF),
```
Comment (whynowy) [inline pkg/sources/source.go:407]: Incorrect Vertex Type.

File: pkg/udf/map_udf.go [hunk 5]
```diff
@@ -239,6 +240,13 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
 
 			// Drop message if it contains the special tag
 			if sharedutil.StringSliceContains(tags, dfv1.MessageTagDrop) {
+				metrics.DroppedMessageCounter.With(map[string]string{
+					metrics.LabelVertex:             vertexName,
+					metrics.LabelPipeline:           pipelineName,
+					metrics.LabelVertexType:         string(dfv1.VertexTypeReduceUDF),
```
Comment (whynowy) [inline pkg/udf/map_udf.go:246]: Same here.

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (whynowy) [inline pkg/metrics/metrics.go:169]: `DroppedMessageCounter` -> `UserDroppedMessages`

File: pkg/sources/source.go [hunk 6]
```diff
@@ -400,6 +401,13 @@ func (sp *SourceProcessor) getTransformerGoWhereDecider(shuffleFuncMap map[strin
 
 		// Drop message if it contains the special tag
 		if sharedutil.StringSliceContains(tags, dfv1.MessageTagDrop) {
+			metrics.UserDroppedMessages.With(map[string]string{
+				metrics.LabelVertex:             sp.VertexInstance.Vertex.Name,
```
Comment (whynowy) [inline pkg/sources/source.go:405]: vertexName should be `sp.VertexInstance.Vertex.Spec.Name`.

**Review by whynowy — APPROVED** [review]
Please re-run `make codegen`.

**Pull request merged**